hal/src/rtl872x/usbd_hid.cpp

```diff
@@ -0,0 +1,379 @@
+
+    return 0;
+}
+
+int HidClassDriver::deinit(unsigned cfgIdx) {
+    if (isConfigured()) {
+        if (txState_) {
+            dev_->flushEndpoint(epIn_);
+        }
+    }
+
+    dev_->closeEndpoint(epIn_);
+
+    epIn_ = ENDPOINT_INVALID;
+    txState_ = false;
+
+    altSetting_ = 0;
+    protocol_ = 0;
+    idleState_ = 0;
+    return 0;
+}
+
+void HidClassDriver::setName(const char* name) {
+    if (name_) {
+        free(name_);
+        name_ = nullptr;
+    }
+    if (name) {
+        // Ignore allocation failures, we'll just default to DEFAULT_NAME
+        name_ = strdup(name);
+    }
+}
+
+int HidClassDriver::setup(SetupRequest* req) {
+    if (req->bmRequestTypeType != SetupRequest::TYPE_CLASS && req->bmRequestTypeType != SetupRequest::TYPE_STANDARD) {
+        return SYSTEM_ERROR_INVALID_ARGUMENT;
+    }
+    if (req->wIndex != interfaceBase_) {
+        return SYSTEM_ERROR_INVALID_ARGUMENT;
+    }
+
+    // Class or standard request
+    if (req->bmRequestTypeType == SetupRequest::TYPE_CLASS) {
+        if (req->bmRequestTypeDirection /* Device to host */) {
+            return handleInSetupRequest(req);
+        } else {
+            return handleOutSetupRequest(req);
+        }
+    } else {
+        // Standard
+        switch (req->bRequest) {
+        case SetupRequest::REQUEST_GET_DESCRIPTOR: {
+            if ((req->wValue >> 8) == hid::DESCRIPTOR_HID_REPORT) {
+                return dev_->setupReply(req, DEFAULT_HID_DESCRIPTOR, reportDescriptorLen());
+            }
+            break;
+        }
+        case SetupRequest::REQUEST_GET_INTERFACE: {
+            return dev_->setupReply(req, &altSetting_, sizeof(altSetting_));
+        }
+        case SetupRequest::REQUEST_SET_INTERFACE: {
+            if (req->wValue == 0) {
+                return dev_->setupReply(req, nullptr, 0);
+            }
+            break;
+        }
+        }
+    }
+    return SYSTEM_ERROR_INVALID_ARGUMENT;
+}
+
+int HidClassDriver::handleInSetupRequest(SetupRequest* req) {
+    switch (req->bRequest) {
+    case hid::GET_PROTOCOL: {
+        return dev_->setupReply(req, (const uint8_t*)&protocol_, sizeof(protocol_));
+    }
+    case hid::GET_IDLE: {
+        return dev_->setupReply(req, (const uint8_t*)&idleState_, sizeof(idleState_));
+    }
+    default: {
+    }
+    }
+    return SYSTEM_ERROR_INVALID_ARGUMENT;
+}
+
+int HidClassDriver::handleOutSetupRequest(SetupRequest* req) {
+    switch (req->bRequest) {
+    case hid::SET_PROTOCOL: {
+        protocol_ = (uint8_t)req->wValue;
+        return 0;
+    }
+    case hid::SET_IDLE: {
+        idleState_ = (uint8_t)req->wValue;
+        return 0;
+    }
+    default: {
+    }
+    }
+    return SYSTEM_ERROR_INVALID_ARGUMENT;
+}
+
+int HidClassDriver::dataIn(unsigned ep, particle::usbd::EndpointEvent ev, size_t len) {
+    if (ep != epIn_) {
+        return SYSTEM_ERROR_UNKNOWN;
+    }
+
+    if (!txState_) {
+        return 0;
```

Comment:
This one should be no-error, as otherwise the rest of the class drivers will be asked to handle this endpoint event, even though we know for a fact that it should have been us.